plan/physical_plan_builder.go

```diff
@@ -43,7 +44,78 @@ const (
 // JoinConcurrency means the number of goroutines that participate in joining.
 var JoinConcurrency = 5
 
-func getRowCountByIndexRanges(sc *variable.StatementContext, table *statistics.Table, indexRanges []*IndexRange, indexInfo *model.IndexInfo) (uint64, error) {
+func getRowCountByIndexRanges(sc *variable.StatementContext, table *statistics.Table, indexRanges []*IndexRange, indexInfo *model.IndexInfo, tableInfo *model.TableInfo) (uint64, error) {
+	var offset int
+	for i := range tableInfo.Indices {
+		if tableInfo.Indices[i].Name.L == indexInfo.Name.L {
+			offset = i
+			break
+		}
+	}
+	if len(table.Indices[offset].Numbers) == 0 {
+		return getRowCountByPseudoIndexRanges(sc, table, indexRanges, indexInfo)
+	}
+	return getRowCountByRealIndexRanges(sc, table, indexRanges, indexInfo, offset)
+}
+
+func getRowCountByRealIndexRanges(sc *variable.StatementContext, table *statistics.Table, indexRanges []*IndexRange, indexInfo *model.IndexInfo, offset int) (uint64, error) {
+	totalCount := int64(0)
+	for _, indexRange := range indexRanges {
+		lv := indexRange.LowVal
+		rv := indexRange.HighVal
+		for i := len(lv); i < len(indexInfo.Columns); i++ {
+			lv = append(lv, types.MinNotNullDatum())
+		}
+		for i := len(rv); i < len(indexInfo.Columns); i++ {
+			rv = append(rv, types.MaxValueDatum())
+		}
+		lb, err := codec.EncodeKey(nil, lv...)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		rb, err := codec.EncodeKey(nil, rv...)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		l := types.NewBytesDatum(lb)
+		r := types.NewBytesDatum(rb)
+		var rowCount int64
+		if lv[0].Kind() == types.KindNull && rv[0].Kind() == types.KindMaxValue {
+			return uint64(table.Count), nil
+		} else if lv[0].Kind() == types.KindMinNotNull {
+			rowCount, err = table.Indices[offset].EqualRowCount(sc, types.Datum{})
```

Comment:
Define a local variable at the beginning of the function like `statsIdx := statsTbl[offset]`.